Implement `sizeOfCoin` and `costOfIncreasingCoin`

Anviking opened the pull request:
Implement and test the following functions:
```haskell
-- | Calculate the cost of increasing a CBOR-encoded Coin-value by another Coin
-- with the lovelace/byte cost given by the 'FeePolicy'.
--
-- Outputs values in the range of [0, 8 * perByteFee]
--
-- >>> let p = FeePolicy (Quantity 0) (Quantity 44)
--
-- >>> costOfIncreasingCoin p 4294967295 1
-- Coin 176 -- (9 bytes - 5 bytes) * 44 lovelace/byte
--
-- >>> costOfIncreasingCoin p 0 4294967296
-- Coin 352 -- 8 bytes * 44 lovelace/byte
costOfIncreasingCoin
    :: FeePolicy
    -> Coin -- ^ Original coin
    -> Coin -- ^ Increment
    -> Coin

-- | Calculate the size of a coin when encoded as CBOR.
sizeOfCoin :: Coin -> TxSize
```

### Future work

The next PR will add a function to `TransactionLayer` with the following shape:
```haskell
    , distributeSurplus
        :: FeePolicy
        -> Coin -- Surplus to distribute
        -> TxFeeAndChange
        -> Either Coin TxFeeAndChange
```

## Issue number

ADP-1514

Anviking committed:
Add sizeOfCoin function

Anviking committed:
Implement costOfIncreasingCoin

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2227,6 +2233,64 @@ balanceTransactionSpec = do
+                    || sizeOfCoin c /= sizeOfCoin (c `Coin.add` (Coin 1))
+
+            it "matches the size of the Word64 CBOR encoding" $
+                property $ checkCoverage $
```
Any objections to hspec-inlined properties?

Anviking committed:
Tweak genEncodingBoundaryLovelace for greater coverage

This improves coverage of boundary cases in sizeOfCoin properties.

jonathanknowles commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -238,38 +239,40 @@ genSlotNo :: Gen SlotNo
+        . fromIntegral
+        . getNonNegative <$> arbitrary @(NonNegative Int) )
+    , (50, Lovelace <$> choose (1_000_000, 1_000_000_000))
+    , (10, Lovelace <$> choose (0, 45_000_000_000_000_000))
```
Suggestion: use `txOutMaxCoin` from `Types.Tx`:

```suggestion
    , (10, Lovelace <$> choose (0, unCoin txOutMaxCoin))
```

Anviking replied:
This would add a dependency on `Cardano.Wallet.Primitive.Types.Tx` however, in an otherwise very standalone module. I'd keep it this way, I think.

jonathanknowles commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -238,38 +239,40 @@ genSlotNo :: Gen SlotNo
+        [ (1, pure 24)
+        , (1, pure 256)
+        , (8, pure 65536)
+        , (90, pure 4294967296)
```
Given that these are powers of two, why not make it explicit:

```suggestion
        , ( 1, pure 24)
        , ( 1, pure $ 2^8)
        , ( 8, pure $ 2^16)
        , (90, pure $ 2^32)
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1436,6 +1437,15 @@ estimateTxCost pp skeleton =
+    | c >= 65536      = TxSize 5
+    | c >= 256        = TxSize 3
+    | c >= 24         = TxSize 2
+    | otherwise       = TxSize 1
```
Suggestion: if we're already using vertical alignment, then perhaps align numerals to the right? (On the basis that it's a more natural alignment for decimal numbers.)

Also: perhaps we could use `NumericUnderscores`.

```suggestion
    | c >= 4_294_967_296 = TxSize 9
    | c >=        65_536 = TxSize 5
    | c >=           256 = TxSize 3
    | c >=            24 = TxSize 2
    | otherwise          = TxSize 1
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2227,6 +2232,50 @@ balanceTransactionSpec = do
+
+        let isBoundary c =
+                sizeOfCoin c /= sizeOfCoin (c `Coin.difference` (Coin 1))
+                || sizeOfCoin c /= sizeOfCoin (c `Coin.add` (Coin 1))
```
Nice boundary check! 👍🏻 

Suggestion: remove extraneous parentheses:

```suggestion
                sizeOfCoin c /= sizeOfCoin (c `Coin.difference` Coin 1)
                || sizeOfCoin c /= sizeOfCoin (c `Coin.add` Coin 1)
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2227,6 +2232,50 @@ balanceTransactionSpec = do
+
+                    -- Use a low coverage requirement of 0.01% just to
+                    -- ensure we see /some/ amount of every size.
+                    let coverSize s = cover 0.01 (s == expected) (show s)
```
I checked what would happen with this very low coverage amount, and it does seem to cause QuickCheck to execute a large number of iterations to verify that the coverage is complete.

However, this property executes very quickly, so it seems fine!

Anviking replied:
> However, this property executes very quickly

Yes! At first I used the `cborSizeOfCoin` above for the real `sizeOfCoin` implementation, but for the sake of making properties faster I decided to implement it with a list of `>=` cases instead.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2227,6 +2232,50 @@ balanceTransactionSpec = do
+                sizeOfCoin (Coin 65536) `shouldBe` TxSize 5
+            it "5 byte to 9 byte boundary" $ do
+                sizeOfCoin (Coin 4294967295) `shouldBe` TxSize 5
+                sizeOfCoin (Coin 4294967296) `shouldBe` TxSize 9
```
Perhaps explicitly indicate that these are powers of 2?

Justification: I suspect it better expresses the intent, and it also emphasises the difference between each pair (which is just `1`), which is harder to see visually if we write out the whole number and need to compare every digit.

```suggestion
            it "2 byte to 3 byte boundary" $ do
                sizeOfCoin (Coin $ 2^8 - 1) `shouldBe` TxSize 2
                sizeOfCoin (Coin $ 2^8    ) `shouldBe` TxSize 3
            it "3 byte to 5 byte boundary" $ do
                sizeOfCoin (Coin $ 2^16 - 1) `shouldBe` TxSize 3
                sizeOfCoin (Coin $ 2^16    ) `shouldBe` TxSize 5
            it "5 byte to 9 byte boundary" $ do
                sizeOfCoin (Coin $ 2^32 - 1) `shouldBe` TxSize 5
                sizeOfCoin (Coin $ 2^32    ) `shouldBe` TxSize 9
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1437,6 +1438,29 @@ estimateTxCost pp skeleton =
+costOfIncreasingCoin (LinearFee fee) from delta =
+    costOfCoin (from <> delta) `Coin.difference` costOfCoin from
+  where
+    perByte = round $ slope fee
```
Suggestion: would using `ceiling` here be more appropriate?

I'm concerned that `round` might lead to underestimation. (But perhaps I'm wrong.)

```suggestion
    perByte = ceiling $ slope fee
```

Anviking replied:
I guess the reason we use a `Double` (we should change this) is that the Byron fee policy coefficient was `Rational`, and not `Natural`. I was here more worried about a `Natural` somehow getting an extra `0.00000000001` — not sure if that's possible — and being incorrectly rounded up.

But I see that we already use [ceiling](https://github.com/input-output-hk/cardano-wallet/blob/db9c4ada4184619c6ff73344d07ffee5e5a44602/lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs#L1395) so I will change.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2276,6 +2277,20 @@ balanceTransactionSpec = do
+        it "costs 176 lovelace to increase 4294967295 by one on mainnet" $ do
+            let feePolicy = LinearFee $ LinearFunction
+                    { intercept = 150_000, slope = 44 }
+            costOfIncreasingCoin feePolicy (Coin 4294967295) (Coin 1)
```
Suggestion: use an explicit power of 2 here.

Justification: it's now really obvious why this is a boundary. (Some readers of the code might not recognize that these values are related to the powers of 2.)

```suggestion
            costOfIncreasingCoin feePolicy (Coin $ 2^32 - 1) (Coin 1)
```

Anviking replied:
https://github.com/input-output-hk/cardano-wallet/pull/3215#issuecomment-1090389142, however because of the test title I think it in particular makes sense to use 4294967295 here. I tweaked the title a bit: 1303e02b1154bb4331c165be2aa6d6cc12bf1862

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2276,6 +2277,20 @@ balanceTransactionSpec = do
+                let feePolicy = LinearFee $ LinearFunction 1 0
+                let res = costOfIncreasingCoin feePolicy c increase
+                counterexample (show res <> "out of bounds") $
+                    res >= (Coin 0) && res <= (Coin 8)
```
I suspect the parentheses are unnecessary:
```suggestion
                    res >= Coin 0 && res <= Coin 8
```

jonathanknowles commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -264,6 +264,9 @@ genEncodingBoundaryLovelace = do
         [ (1, choose (-10, 10))
 
+        -- Either offset by -1 (just below boundary), or 0 (just above boundary)
+        , (1, choose (-1, 0))
```
Question: what is the benefit of choosing `0` here?

Anviking replied:
That we make sure to generate e.g. `2^32`, which is encoded as 9 bytes, not just `2^32 - 1` which is 5 bytes.

jonathanknowles reviewed (commented):
Hi @Anviking 

I've made some general comments and suggestions. Will have another look in a bit!

Anviking committed:
fixup: remove redundant parenthesis

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
fixup: use ceiling rather than round for fee

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
Use NumericUnderscores and right-alignment in sizeOfCoin

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
fixup: adjust wording in costOfIncreasingCoin test

jonathanknowles committed:
Correct indentation for exports of `Cardano.Api.Gen`.

Use 4-space indents, and put exports into alphabetical order.

jonathanknowles committed:
Introduce the named constants `txOut{Min,Max}Lovelace`.

This commit also uses these named constants in `genLovelace` to replace
hard-coded values.

If a developer is new to the codebase, they can understand the choice of
these constants by navigating to their definitions and reading the
comments.

jonathanknowles committed:
Remove now-redundant use of `Lovelace` constructor.

Since `Lovelace` is an instance of `Num` and `Random`, we can avoid
wrapping and unwrapping `Lovelace` values when using `choose`.

jonathanknowles committed:
Use statically-checked `intCast` to replace `fromIntegral`.

jonathanknowles committed:
Simplify generator by moving the location of a type application.

jonathanknowles committed:
Use `NumericUnderscores` and add comments to inline power-of-two constants.

jonathanknowles committed:
Add comments to `sizeOfCoin` explaining the choices of constants.

jonathanknowles committed:
Use explicit powers-of-two in `sizeOfCoin` boundary tests.

In the implementation of `sizeOfCoin`, we use hard-coded constants for
powers of two, which avoids computing these values on each call.

However, in the test code, we can be even clearer about our intent, by
explicitly stating which powers of two we're interested in.

For someone new to the codebase, being able to see the relationship with
the powers of two is arguably helpful for being able to comprehend why
we're testing these particular values.

At a glance, it's arguably quite hard to see the the difference between:
    - 4294967295
    - 4294967296

But it's very easy to see the difference between:
    - 2^32 - 1
    - 2^32

jonathanknowles committed:
Use explicit powers-of-two in test for `costOfIncreasingCoin`.

In this commit, we also remove the expected cost increase from the name
of the test. Otherwise, if someone updates this constant in future, it's
possible that they'll forget to update the name of the test case, and
the compiler won't tell them.

jonathanknowles committed:
Add monomorphic function `power`.

Anviking committed:
Remove redundant comment in sizeOfCoin

Anviking committed:
Tweak title of costOfIncreasingCoin test

- Lead with "[it] costs"
- Clearly specify the ≈4.29 ada value in the title

jonathanknowles reviewed (approved):
Nice work! 👍🏻
